integration/benchmark/cmd/ci/main.go

```diff
@@ -163,60 +170,89 @@ func main() {
+		osVersion)
+
+	select {
+	case <-time.After(loadCase.duration):
+		lg.Stop()
+	case <-ctx.Done(): //TODO: the loader currently doesn't ever cancel the context
+		lg.Stop()
+		// add logging here to express the *why* of lg choose to cancel the context.
+		// when the loader cancels its own context it may also call Stop() on itself
+		// this may become redundant.
+	}
+
 	prepareDataForBigQuery(dataSlices, *ciFlag)
 }
 
-func calculateTpsSlices(start, end time.Time, leadTime, sliceTime, commit, goVersion, osVersion string, lg *benchmark.ContLoadGenerator) []dataSlice {
-	//remove the lead time on both start and end, this should remove spin-up and spin-down times
-	leadDuration, _ := time.ParseDuration(leadTime)
-	endTime := end.Add(-1 * leadDuration)
-	sliceDuration, _ := time.ParseDuration(sliceTime)
+func recordTransactionData(
+	slices chan dataSlice,
+	lg *benchmark.ContLoadGenerator,
+	sliceDuration time.Duration,
+	runStartTime time.Time,
+	gitSha, goVersion, osVersion string) {
+	// get initial values for first slice
+	startTime := time.Now()
+	startExecutedTransactions := lg.GetTxExecuted()
+
+	t := time.NewTicker(sliceDuration)
+	defer t.Stop()
+
+	for {
+		select {
+		case <-t.C:
+			endTime := time.Now()
+			endExecutedTransaction := lg.GetTxExecuted()
+
+			// calculate this slice
+			inputTps := lg.AvgTpsBetween(startTime, endTime)
+			outputTps := float64(endExecutedTransaction-startExecutedTransactions) / sliceDuration.Seconds()
+			slice := dataSlice{
+				GitSha:       gitSha,
+				GoVersion:    goVersion,
+				OsVersion:    osVersion,
+				StartTime:    startTime,
+				EndTime:      endTime,
+				InputTps:     inputTps,
+				OutputTps:    outputTps,
+				BeforeExTxs:  startExecutedTransactions,
+				AfterExTxs:   endExecutedTransaction,
+				RunStartTime: runStartTime}
+
+			// set start values for next slice
+			startExecutedTransactions = endExecutedTransaction
+			startTime = endTime
+
+			slices <- slice
+		case <-lg.Done():
+			close(slices)
+			return
+		}
+	}
+}
 
+func prepareDataForBigQuery(slicesChannel chan dataSlice, ci bool) {
 	slices := make([]dataSlice, 0)
 
-	for currentTime := start.Add(leadDuration); currentTime.Add(sliceDuration).Before(endTime); currentTime = currentTime.Add(sliceDuration) {
-		sliceEndTime := currentTime.Add(sliceDuration)
-
-		inputTps := lg.AvgTpsBetween(currentTime, sliceEndTime)
-		proStart := getPrometheusTransactionAtTime(currentTime)
-		proEnd := getPrometheusTransactionAtTime(sliceEndTime)
-
-		outputTps := (proEnd - proStart) / (sliceDuration).Seconds()
-
-		slice := dataSlice{
-			GitSha:              commit,
-			GoVersion:           goVersion,
-			OsVersion:           osVersion,
-			StartTime:           currentTime,
-			EndTime:             currentTime.Add(sliceDuration),
-			InputTps:            inputTps,
-			OutputTps:           outputTps,
-			ProStartTransaction: proStart,
-			ProEndTransaction:   proEnd}
-
-		slices = append(slices, slice)
+	for len(slicesChannel) != 0 {
```

Comment:
I think you can just iterate over the channel

```
for slice := slicesChannel {
  slices = append(slices, slice)
}
```

Reply:
works with `for slice := range slicesChannel` 

will update